controllers/artefacts.rb

```diff
@@ -183,8 +202,70 @@ def get_ontology_and_latest_submission
                 @latest_submission.bring(ontology: [:acronym])
                 return @ontology, @latest_submission
             end
+            
+            def fetch_resource
+                uri = params['uri']
+                ontology_acronym = params['artefactID']
+                
+                error 404, "The uri parameter must be provided via ?uri=<uri>" if uri.nil?
+              
+                ontology, latest_submission = get_ontology_and_latest_submission
+                check_access(ontology)
+              
+                fq = [
+                  "ontology_t:\"#{ontology_acronym}\"",
+                  "resource_id:\"#{uri}\""
+                ]
+                conn = SOLR::SolrConnector.new(Goo.search_conf, :ontology_data)
+                resp = conn.search("*:*", fq: fq, defType: "edismax", start: 0, rows: 1)
+                doc = resp["response"]["docs"].first
+
+                if doc
+                  type = doc["type_t"] || doc["type_txt"].first
+              
+                  resource = case type
```

Comment:
this is same at line 150, you can create a helper called `model_from_type(type_str)` and return a Model.